{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CCDS Tagging Target Search\n",
    "\n",
    "## Overview:\n",
    "- Intended to search for relevant targets in human exons given parameters about target site and position\n",
    "- Reads in nucleotide sequences of human genes from the CCDS files of the human genome\n",
    "- Current strategy for capturing 3' UTRs:\n",
    "    - Search for TTAC sequences near the 3' end of genes\n",
    "    - From CCDS IDs, look up ENST IDs using ENSEMBL Biomart\n",
    "    - Find relevant ENST IDs in gencode.v38.pc_transcripts\n",
    "    - Check for PAM etc.\n",
    "\n",
    "## To Do:\n",
    "- Find a different input file or modify current for better analysis of sense strand targets\n",
    "    - Sequences all end at the stop codon of the gene\n",
    "    - This prevents a good search of sense strand TTACs very close to the end as the spacer sequence would be further downstream... would want 3' UTRs for all of these genes\n",
    "- Search in the RC direction\n",
    "    - Trivial to implement computationally, unclear if relevant biologically\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import generally useful packages\n",
    "import csv\n",
    "import itertools\n",
    "import sys\n",
    "import os\n",
    "import math\n",
    "import numpy as np\n",
    "from tqdm.notebook import tnrange, tqdm_notebook, tqdm\n",
    "\n",
    "#Packages for data analysis\n",
    "import pandas as pd\n",
    "\n",
    "#Packages for sequence parsing\n",
    "import regex as re\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chromosome</th>\n",
       "      <th>nc_accession</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>ccds_status</th>\n",
       "      <th>cds_strand</th>\n",
       "      <th>cds_from</th>\n",
       "      <th>cds_to</th>\n",
       "      <th>cds_locations</th>\n",
       "      <th>match_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>NC_000001.8</td>\n",
       "      <td>LINC00115</td>\n",
       "      <td>79854</td>\n",
       "      <td>CCDS1.1</td>\n",
       "      <td>Withdrawn</td>\n",
       "      <td>-</td>\n",
       "      <td>801942</td>\n",
       "      <td>802433</td>\n",
       "      <td>[801942-802433]</td>\n",
       "      <td>Identical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>148398</td>\n",
       "      <td>CCDS2.2</td>\n",
       "      <td>Public</td>\n",
       "      <td>+</td>\n",
       "      <td>925941</td>\n",
       "      <td>944152</td>\n",
       "      <td>[925941-926012, 930154-930335, 931038-931088, ...</td>\n",
       "      <td>Identical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>26155</td>\n",
       "      <td>CCDS3.1</td>\n",
       "      <td>Public</td>\n",
       "      <td>-</td>\n",
       "      <td>944693</td>\n",
       "      <td>959239</td>\n",
       "      <td>[944693-944799, 945056-945145, 945517-945652, ...</td>\n",
       "      <td>Identical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>84069</td>\n",
       "      <td>CCDS4.1</td>\n",
       "      <td>Public</td>\n",
       "      <td>+</td>\n",
       "      <td>966531</td>\n",
       "      <td>974574</td>\n",
       "      <td>[966531-966613, 966703-966802, 970276-970422, ...</td>\n",
       "      <td>Identical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>HES4</td>\n",
       "      <td>57801</td>\n",
       "      <td>CCDS5.1</td>\n",
       "      <td>Public</td>\n",
       "      <td>-</td>\n",
       "      <td>999058</td>\n",
       "      <td>999972</td>\n",
       "      <td>[999058-999431, 999525-999612, 999691-999786, ...</td>\n",
       "      <td>Identical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35133</th>\n",
       "      <td>X</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>HSFX4</td>\n",
       "      <td>101927685</td>\n",
       "      <td>CCDS87786.1</td>\n",
       "      <td>Public</td>\n",
       "      <td>+</td>\n",
       "      <td>149929644</td>\n",
       "      <td>149931104</td>\n",
       "      <td>[149929644-149930126, 149930586-149931104]</td>\n",
       "      <td>Identical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35134</th>\n",
       "      <td>X</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>VMA21</td>\n",
       "      <td>203547</td>\n",
       "      <td>CCDS87789.1</td>\n",
       "      <td>Public</td>\n",
       "      <td>+</td>\n",
       "      <td>151396839</td>\n",
       "      <td>151405057</td>\n",
       "      <td>[151396839-151397056, 151403630-151403739, 151...</td>\n",
       "      <td>Identical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35135</th>\n",
       "      <td>X</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>PNMA6F</td>\n",
       "      <td>105373377</td>\n",
       "      <td>CCDS87792.1</td>\n",
       "      <td>Public</td>\n",
       "      <td>-</td>\n",
       "      <td>153318937</td>\n",
       "      <td>153320673</td>\n",
       "      <td>[153318937-153320673]</td>\n",
       "      <td>Identical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35136</th>\n",
       "      <td>X</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>PNMA6E</td>\n",
       "      <td>649238</td>\n",
       "      <td>CCDS87793.1</td>\n",
       "      <td>Public</td>\n",
       "      <td>-</td>\n",
       "      <td>153396905</td>\n",
       "      <td>153398848</td>\n",
       "      <td>[153396905-153397871, 153398706-153398848]</td>\n",
       "      <td>Identical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35137</th>\n",
       "      <td>X</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>FAM3A</td>\n",
       "      <td>60343</td>\n",
       "      <td>CCDS87800.1</td>\n",
       "      <td>Public</td>\n",
       "      <td>-</td>\n",
       "      <td>154506810</td>\n",
       "      <td>154515771</td>\n",
       "      <td>[154506810-154506905, 154507202-154507328, 154...</td>\n",
       "      <td>Identical</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35138 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      #chromosome  nc_accession       gene    gene_id      ccds_id  \\\n",
       "0               1   NC_000001.8  LINC00115      79854      CCDS1.1   \n",
       "1               1  NC_000001.11     SAMD11     148398      CCDS2.2   \n",
       "2               1  NC_000001.11      NOC2L      26155      CCDS3.1   \n",
       "3               1  NC_000001.11    PLEKHN1      84069      CCDS4.1   \n",
       "4               1  NC_000001.11       HES4      57801      CCDS5.1   \n",
       "...           ...           ...        ...        ...          ...   \n",
       "35133           X  NC_000023.11      HSFX4  101927685  CCDS87786.1   \n",
       "35134           X  NC_000023.11      VMA21     203547  CCDS87789.1   \n",
       "35135           X  NC_000023.11     PNMA6F  105373377  CCDS87792.1   \n",
       "35136           X  NC_000023.11     PNMA6E     649238  CCDS87793.1   \n",
       "35137           X  NC_000023.11      FAM3A      60343  CCDS87800.1   \n",
       "\n",
       "      ccds_status cds_strand   cds_from     cds_to  \\\n",
       "0       Withdrawn          -     801942     802433   \n",
       "1          Public          +     925941     944152   \n",
       "2          Public          -     944693     959239   \n",
       "3          Public          +     966531     974574   \n",
       "4          Public          -     999058     999972   \n",
       "...           ...        ...        ...        ...   \n",
       "35133      Public          +  149929644  149931104   \n",
       "35134      Public          +  151396839  151405057   \n",
       "35135      Public          -  153318937  153320673   \n",
       "35136      Public          -  153396905  153398848   \n",
       "35137      Public          -  154506810  154515771   \n",
       "\n",
       "                                           cds_locations match_type  \n",
       "0                                        [801942-802433]  Identical  \n",
       "1      [925941-926012, 930154-930335, 931038-931088, ...  Identical  \n",
       "2      [944693-944799, 945056-945145, 945517-945652, ...  Identical  \n",
       "3      [966531-966613, 966703-966802, 970276-970422, ...  Identical  \n",
       "4      [999058-999431, 999525-999612, 999691-999786, ...  Identical  \n",
       "...                                                  ...        ...  \n",
       "35133         [149929644-149930126, 149930586-149931104]  Identical  \n",
       "35134  [151396839-151397056, 151403630-151403739, 151...  Identical  \n",
       "35135                              [153318937-153320673]  Identical  \n",
       "35136         [153396905-153397871, 153398706-153398848]  Identical  \n",
       "35137  [154506810-154506905, 154507202-154507328, 154...  Identical  \n",
       "\n",
       "[35138 rows x 11 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Pre-process CCDS key into dictionary format to speed up name look-up\n",
    "genes = pd.read_csv('CCDS.20180614.csv')\n",
    "gene_dict = {}\n",
    "\n",
    "for index, gene in genes.iterrows():\n",
    "    gene_dict[gene[\"ccds_id\"]]=gene[\"gene\"]\n",
    "    \n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('ATGTCCAAGGGGATCCTGCAGGTGCATCCTCCGATCTGCGACTGCCCGGGCTGC...TGA', SingleLetterAlphabet()), id='CCDS2.2|Hs109|chr1', name='CCDS2.2|Hs109|chr1', description='CCDS2.2|Hs109|chr1', dbxrefs=[])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Examine file structure of parsed sequences from CCDS\n",
    "genesequences = SeqIO.parse(\"CCDS_nucleotide.current.fna\", \"fasta\")\n",
    "next(genesequences)\n",
    "\n",
    "#Note that each sequence has a name/id, and ends with stop codon of gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2c1b064cf3314120bf83114c021af4a1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=1.0, bar_style='info', max=1.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of matching sequences: 150\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CCDS Name</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>gRNA</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>TTAC position</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Distance from 3' End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCDS746.1|Hs109|chr1</td>\n",
       "      <td>GCLM</td>\n",
       "      <td>TACATTTTACAAGCTAAAAGAAGGGGT</td>\n",
       "      <td>ATGGGCACCGACAGCCGCGCGGCCAAGGCGCTCCTGGCGCGGGCCC...</td>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCDS1037.1|Hs109|chr1</td>\n",
       "      <td>S100A12</td>\n",
       "      <td>CTGCCCATTACCACACCCACAAAGAGT</td>\n",
       "      <td>ATGACAAAACTTGAAGAGCATCTGGAGGGAATTGTCAATATCTTCC...</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCDS1285.1|Hs109|chr1</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>TAAAACGTTACATACATACTCTAGGGT</td>\n",
       "      <td>ATGTTTTTACCTCATATGAACCACCTGACATTGGAACAGACTTTCT...</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CCDS1288.1|Hs109|chr1</td>\n",
       "      <td>KIFAP3</td>\n",
       "      <td>ATGAACCTTACTACTATGGCTATGGAT</td>\n",
       "      <td>ATGCAAGGGGAGGACGCCAGATACCTCAAAAGGAAAGTTAAAGGAG...</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CCDS1362.1|Hs109|chr1</td>\n",
       "      <td>C1orf21</td>\n",
       "      <td>CGGGATTACTGTTCGGAAGAAGAGGAT</td>\n",
       "      <td>ATGGGCTGTGCCTCCGCCAAGCATGTTGCCACTGTTCAAAATGAAG...</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>CCDS86187.1|Hs109|chr11</td>\n",
       "      <td>LGR4</td>\n",
       "      <td>GCGCTATGCTTACAATCTACCAAGAGT</td>\n",
       "      <td>ATGCCGGGCCCGCTAGGGCTGCTCTGCTTCCTCGCCCTGGGGCTGC...</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>CCDS86584.1|Hs109|chr17</td>\n",
       "      <td>B9D1</td>\n",
       "      <td>CATACAGGTTACAGCCTGGGAGTGGGT</td>\n",
       "      <td>ATGGCGACCGCGAGTCCTAGCGTCTTTCTACTCATGGTCAACGGGC...</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>CCDS86588.1|Hs109|chr17</td>\n",
       "      <td>ADAP2</td>\n",
       "      <td>AACCGGCTTACTGCATCAACAGAGAGT</td>\n",
       "      <td>ATGGGCGATCGCGAGCGCAACAAGAAGCGGCTGCTGGAGCTGCTGC...</td>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>CCDS87280.1|Hs109|chr4</td>\n",
       "      <td>HPGD</td>\n",
       "      <td>CCCAGGGTTACTTTAGAACTGCTGGAT</td>\n",
       "      <td>ATGCACGTGAACGGCAAAGTGGCGCTGGTGACCGGCGCGGCTCAGG...</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>CCDS87676.1|Hs109|chr9</td>\n",
       "      <td>INIP</td>\n",
       "      <td>CTGTTTTACCTCGCCTTGACCCAGAAT</td>\n",
       "      <td>ATGCAGAACCAGTCTTCAACAAATCATCCTGGAGCTAGCATTGCAC...</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>150 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   CCDS Name Gene Name                         gRNA  \\\n",
       "0       CCDS746.1|Hs109|chr1      GCLM  TACATTTTACAAGCTAAAAGAAGGGGT   \n",
       "1      CCDS1037.1|Hs109|chr1   S100A12  CTGCCCATTACCACACCCACAAAGAGT   \n",
       "2      CCDS1285.1|Hs109|chr1  C1orf112  TAAAACGTTACATACATACTCTAGGGT   \n",
       "3      CCDS1288.1|Hs109|chr1    KIFAP3  ATGAACCTTACTACTATGGCTATGGAT   \n",
       "4      CCDS1362.1|Hs109|chr1   C1orf21  CGGGATTACTGTTCGGAAGAAGAGGAT   \n",
       "..                       ...       ...                          ...   \n",
       "145  CCDS86187.1|Hs109|chr11      LGR4  GCGCTATGCTTACAATCTACCAAGAGT   \n",
       "146  CCDS86584.1|Hs109|chr17      B9D1  CATACAGGTTACAGCCTGGGAGTGGGT   \n",
       "147  CCDS86588.1|Hs109|chr17     ADAP2  AACCGGCTTACTGCATCAACAGAGAGT   \n",
       "148   CCDS87280.1|Hs109|chr4      HPGD  CCCAGGGTTACTTTAGAACTGCTGGAT   \n",
       "149   CCDS87676.1|Hs109|chr9      INIP  CTGTTTTACCTCGCCTTGACCCAGAAT   \n",
       "\n",
       "                                              Sequence  TTAC position  Frame  \\\n",
       "0    ATGGGCACCGACAGCCGCGCGGCCAAGGCGCTCCTGGCGCGGGCCC...             15      2   \n",
       "1    ATGACAAAACTTGAAGAGCATCTGGAGGGAATTGTCAATATCTTCC...             14      1   \n",
       "2    ATGTTTTTACCTCATATGAACCACCTGACATTGGAACAGACTTTCT...             14      1   \n",
       "3    ATGCAAGGGGAGGACGCCAGATACCTCAAAAGGAAAGTTAAAGGAG...             14      1   \n",
       "4    ATGGGCTGTGCCTCCGCCAAGCATGTTGCCACTGTTCAAAATGAAG...             16      1   \n",
       "..                                                 ...            ...    ...   \n",
       "145  ATGCCGGGCCCGCTAGGGCTGCTCTGCTTCCTCGCCCTGGGGCTGC...             12      1   \n",
       "146  ATGGCGACCGCGAGTCCTAGCGTCTTTCTACTCATGGTCAACGGGC...             13      3   \n",
       "147  ATGGGCGATCGCGAGCGCAACAAGAAGCGGCTGCTGGAGCTGCTGC...             14      3   \n",
       "148  ATGCACGTGAACGGCAAAGTGGCGCTGGTGACCGGCGCGGCTCAGG...             14      1   \n",
       "149  ATGCAGAACCAGTCTTCAACAAATCATCCTGGAGCTAGCATTGCAC...             16      2   \n",
       "\n",
       "     Distance from 3' End  \n",
       "0                      23  \n",
       "1                      18  \n",
       "2                      18  \n",
       "3                      21  \n",
       "4                      27  \n",
       "..                    ...  \n",
       "145                    24  \n",
       "146                    34  \n",
       "147                    34  \n",
       "148                    30  \n",
       "149                    20  \n",
       "\n",
       "[150 rows x 7 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#SaCas9 search in the forward direction, TTACs in window of 10-17, any frame\n",
    "\n",
    "# Parameters:\n",
    "search_regex = \"TTAC\\w{8,15}G[G|A][G|A]T\" #Regex for sequence to match\n",
    "distance_from_end = 45 #Maximum distance from 3' end of gene to TTAC = last 20 aas\n",
    "gRNA_length = 27 #Includes PAM\n",
    "PAM_length = 6\n",
    "file_name = \"SaCas9_10-17_Forward_15aa\"\n",
    "\n",
    "'''\n",
    "SaCas9 has a gRNA structure of GN(21)NNGRRT\n",
    "PAM regex is thus \"\\w{2}G(G|A)(G|A)T\"\n",
    "Want TTAC from window of 10-17 -> \"TTAC\\w{8-15}G[G|A][G|A]T\"\n",
    "'''\n",
    "\n",
    "#Initialize relevant variables\n",
    "counter = 0\n",
    "good_sequences = []\n",
    "filtered_data = []\n",
    "\n",
    "#Load file\n",
    "genesequences = SeqIO.parse(\"CCDS_nucleotide.current.fna\", \"fasta\")\n",
    "pbar = tqdm(genesequences)\n",
    "\n",
    "#Iterate through sequences and search for matches\n",
    "for sequence in pbar: \n",
    "    target_RC_search = re.finditer(search_regex, str(sequence.seq)) #Regex match iterator\n",
    "    for match in target_RC_search:\n",
    "        (start, stop) = match.span()\n",
    "        real_start = start-(gRNA_length-len(match.group())) #Compensates for parts of gRNA missing from match\n",
    "        if real_start+1 >= len(sequence)-distance_from_end:\n",
    "            TTAC_pos = len(match.group())-PAM_length\n",
    "            frame = (start+4)%3 + 1\n",
    "            gRNA_dist = len(sequence)-start-4\n",
    "            gRNA_sequence = str(sequence.seq)[real_start:real_start+gRNA_length]\n",
    "            gene_name = gene_dict[re.match(\"CCDS.*\\.\\d\", sequence.name).group()]\n",
    "            good_sequences.append(sequence)\n",
    "            filtered_data.append([sequence.name, gene_name, gRNA_sequence, str(sequence.seq), TTAC_pos, frame, gRNA_dist])\n",
    "            counter += 1\n",
    "    \n",
    "print(f\"Number of matching sequences: {counter}\")\n",
    "\n",
    "#Write data to relevant files\n",
    "with open(file_name+\"_Matches.fasta\", \"w\") as output_handle:\n",
    "        SeqIO.write(good_sequences, output_handle, \"fasta\")\n",
    "        \n",
    "df = pd.DataFrame(filtered_data, columns = [\"CCDS Name\", \"Gene Name\", \"gRNA\", \"Sequence\", \"TTAC position\", \"Frame\", \"Distance from 3' End\"])\n",
    "df.to_csv(file_name+\".csv\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "310ab0d30f2341de8197aa2f7d3c23c5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=1.0, bar_style='info', max=1.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of matching sequences: 290\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CCDS Name</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>gRNA</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>TTAC position</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Distance from 3' End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCDS627.1|Hs109|chr1</td>\n",
       "      <td>UBE2U</td>\n",
       "      <td>CAGTATTACAAATGGAAGAAAATGGAT</td>\n",
       "      <td>ATGCACGGCAGAGCTTACCTCTTGCTGCACAGAGACTTCTGTGATC...</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCDS694.1|Hs109|chr1</td>\n",
       "      <td>DNASE2B</td>\n",
       "      <td>GGCAAATTTACCAAGCATTTCAAGGAT</td>\n",
       "      <td>ATGCCCCAGCTGTGCACCAGGGCCAGCTCATCAGAGATTCCTGGCA...</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCDS746.1|Hs109|chr1</td>\n",
       "      <td>GCLM</td>\n",
       "      <td>TACATTTTACAAGCTAAAAGAAGGGGT</td>\n",
       "      <td>ATGGGCACCGACAGCCGCGCGGCCAAGGCGCTCCTGGCGCGGGCCC...</td>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CCDS867.1|Hs109|chr1</td>\n",
       "      <td>HIPK1</td>\n",
       "      <td>ACAATTTACACTGGATACCCGCTGAGT</td>\n",
       "      <td>ATGGCATCACAGCTGCAAGTGTTTTCGCCCCCATCAGTGTCGTCGA...</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CCDS869.1|Hs109|chr1</td>\n",
       "      <td>HIPK1</td>\n",
       "      <td>ACAATTTACACTGGATACCCGCTGAGT</td>\n",
       "      <td>ATGGTTTTGATGTTTCAGATTCGTTATATTTCACAAACACAAGGCT...</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>CCDS86966.1|Hs109|chr20</td>\n",
       "      <td>OSBPL2</td>\n",
       "      <td>TATGCAGGGGATTACTTTGAGCGGAAT</td>\n",
       "      <td>ATGCCAATCGCCTTCAACGAGCCTCTGAGCTTCTTGCAGCGGATCA...</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>CCDS87159.1|Hs109|chr3</td>\n",
       "      <td>SLC33A1</td>\n",
       "      <td>AATTTAAAAAGTTACAGGATGAAGGAT</td>\n",
       "      <td>ATGTCACCCACCATCTCCCACAAGGACAGCAGCCGGCAACGGCGGC...</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>CCDS87280.1|Hs109|chr4</td>\n",
       "      <td>HPGD</td>\n",
       "      <td>CCCAGGGTTACTTTAGAACTGCTGGAT</td>\n",
       "      <td>ATGCACGTGAACGGCAAAGTGGCGCTGGTGACCGGCGCGGCTCAGG...</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>CCDS87482.1|Hs109|chr7</td>\n",
       "      <td>BZW2</td>\n",
       "      <td>AGTGGTTACAAAATGCAGAAGAAGAAT</td>\n",
       "      <td>ATGGCAGAAAAAGATGCCAACTCTGTTACCTCGTCTTTGAGAAAAG...</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>CCDS87676.1|Hs109|chr9</td>\n",
       "      <td>INIP</td>\n",
       "      <td>CTGTTTTACCTCGCCTTGACCCAGAAT</td>\n",
       "      <td>ATGCAGAACCAGTCTTCAACAAATCATCCTGGAGCTAGCATTGCAC...</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>290 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   CCDS Name Gene Name                         gRNA  \\\n",
       "0       CCDS627.1|Hs109|chr1     UBE2U  CAGTATTACAAATGGAAGAAAATGGAT   \n",
       "1       CCDS694.1|Hs109|chr1   DNASE2B  GGCAAATTTACCAAGCATTTCAAGGAT   \n",
       "2       CCDS746.1|Hs109|chr1      GCLM  TACATTTTACAAGCTAAAAGAAGGGGT   \n",
       "3       CCDS867.1|Hs109|chr1     HIPK1  ACAATTTACACTGGATACCCGCTGAGT   \n",
       "4       CCDS869.1|Hs109|chr1     HIPK1  ACAATTTACACTGGATACCCGCTGAGT   \n",
       "..                       ...       ...                          ...   \n",
       "285  CCDS86966.1|Hs109|chr20    OSBPL2  TATGCAGGGGATTACTTTGAGCGGAAT   \n",
       "286   CCDS87159.1|Hs109|chr3   SLC33A1  AATTTAAAAAGTTACAGGATGAAGGAT   \n",
       "287   CCDS87280.1|Hs109|chr4      HPGD  CCCAGGGTTACTTTAGAACTGCTGGAT   \n",
       "288   CCDS87482.1|Hs109|chr7      BZW2  AGTGGTTACAAAATGCAGAAGAAGAAT   \n",
       "289   CCDS87676.1|Hs109|chr9      INIP  CTGTTTTACCTCGCCTTGACCCAGAAT   \n",
       "\n",
       "                                              Sequence  TTAC position  Frame  \\\n",
       "0    ATGCACGGCAGAGCTTACCTCTTGCTGCACAGAGACTTCTGTGATC...             16      1   \n",
       "1    ATGCCCCAGCTGTGCACCAGGGCCAGCTCATCAGAGATTCCTGGCA...             14      1   \n",
       "2    ATGGGCACCGACAGCCGCGCGGCCAAGGCGCTCCTGGCGCGGGCCC...             15      2   \n",
       "3    ATGGCATCACAGCTGCAAGTGTTTTCGCCCCCATCAGTGTCGTCGA...             16      1   \n",
       "4    ATGGTTTTGATGTTTCAGATTCGTTATATTTCACAAACACAAGGCT...             16      1   \n",
       "..                                                 ...            ...    ...   \n",
       "285  ATGCCAATCGCCTTCAACGAGCCTCTGAGCTTCTTGCAGCGGATCA...             10      1   \n",
       "286  ATGTCACCCACCATCTCCCACAAGGACAGCAGCCGGCAACGGCGGC...             10      2   \n",
       "287  ATGCACGTGAACGGCAAAGTGGCGCTGGTGACCGGCGCGGCTCAGG...             14      1   \n",
       "288  ATGGCAGAAAAAGATGCCAACTCTGTTACCTCGTCTTTGAGAAAAG...             16      2   \n",
       "289  ATGCAGAACCAGTCTTCAACAAATCATCCTGGAGCTAGCATTGCAC...             16      2   \n",
       "\n",
       "     Distance from 3' End  \n",
       "0                      51  \n",
       "1                      45  \n",
       "2                      23  \n",
       "3                      51  \n",
       "4                      51  \n",
       "..                    ...  \n",
       "285                    39  \n",
       "286                    44  \n",
       "287                    30  \n",
       "288                    44  \n",
       "289                    20  \n",
       "\n",
       "[290 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#SaCas9 search in the forward direction, TTACs in window of 10-17, any frame\n",
    "\n",
    "# Parameters:\n",
    "search_regex = \"TTAC\\w{8,15}G[G|A][G|A]T\" #Regex for sequence to match\n",
    "distance_from_end = 60 #Maximum distance from 3' end of gene to TTAC = last 20 aas\n",
    "gRNA_length = 27 #Includes PAM\n",
    "PAM_length = 6\n",
    "file_name = \"SaCas9_10-17_Forward_20aa\"\n",
    "\n",
    "'''\n",
    "SaCas9 has a gRNA structure of GN(21)NNGRRT\n",
    "PAM regex is thus \"\\w{2}G(G|A)(G|A)T\"\n",
    "Want TTAC from window of 10-17 -> \"TTAC\\w{8-15}G[G|A][G|A]T\"\n",
    "'''\n",
    "\n",
    "#Initialize relevant variables\n",
    "counter = 0\n",
    "good_sequences = []\n",
    "filtered_data = []\n",
    "\n",
    "#Load file\n",
    "genesequences = SeqIO.parse(\"CCDS_nucleotide.current.fna\", \"fasta\")\n",
    "pbar = tqdm(genesequences)\n",
    "\n",
    "#Iterate through sequences and search for matches\n",
    "for sequence in pbar: \n",
    "    target_RC_search = re.finditer(search_regex, str(sequence.seq)) #Regex match iterator\n",
    "    for match in target_RC_search:\n",
    "        (start, stop) = match.span()\n",
    "        real_start = start-(gRNA_length-len(match.group())) #Compensates for parts of gRNA missing from match\n",
    "        if real_start+1 >= len(sequence)-distance_from_end:\n",
    "            TTAC_pos = len(match.group())-PAM_length\n",
    "            frame = (start+4)%3 + 1\n",
    "            gRNA_dist = len(sequence)-start-4\n",
    "            gRNA_sequence = str(sequence.seq)[real_start:real_start+gRNA_length]\n",
    "            gene_name = gene_dict[re.match(\"CCDS.*\\.\\d\", sequence.name).group()]\n",
    "            good_sequences.append(sequence)\n",
    "            filtered_data.append([sequence.name, gene_name, gRNA_sequence, str(sequence.seq), TTAC_pos, frame, gRNA_dist])\n",
    "            counter += 1\n",
    "    \n",
    "print(f\"Number of matching sequences: {counter}\")\n",
    "\n",
    "#Write data to relevant files\n",
    "with open(file_name+\"_Matches.fasta\", \"w\") as output_handle:\n",
    "        SeqIO.write(good_sequences, output_handle, \"fasta\")\n",
    "        \n",
    "df = pd.DataFrame(filtered_data, columns = [\"CCDS Name\", \"Gene Name\", \"gRNA\", \"Sequence\", \"TTAC position\", \"Frame\", \"Distance from 3' End\"])\n",
    "df.to_csv(file_name+\".csv\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a1d9a38414c1431c839c9e9c4183c0df",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=1.0, bar_style='info', max=1.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of matching sequences: 354\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CCDS Name</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>gRNA</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>TTAC position</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Distance from 3' End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCDS157.1|Hs109|chr1</td>\n",
       "      <td>CELA2A</td>\n",
       "      <td>GCGGGTCTCCAATTACATCGACTGGAT</td>\n",
       "      <td>ATGATAAGGACGCTGCTGCTGTCCACTTTGGTGGCTGGAGCCCTCA...</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCDS450.1|Hs109|chr1</td>\n",
       "      <td>COL9A2</td>\n",
       "      <td>CCTCTGCCCGCCTTACAGAGCCTGGAT</td>\n",
       "      <td>ATGGCCGCCGCTACGGCCTCCCCCCGCAGCCTCCTTGTTCTCCTCC...</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCDS627.1|Hs109|chr1</td>\n",
       "      <td>UBE2U</td>\n",
       "      <td>CAGTATTACAAATGGAAGAAAATGGAT</td>\n",
       "      <td>ATGCACGGCAGAGCTTACCTCTTGCTGCACAGAGACTTCTGTGATC...</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CCDS694.1|Hs109|chr1</td>\n",
       "      <td>DNASE2B</td>\n",
       "      <td>GGCAAATTTACCAAGCATTTCAAGGAT</td>\n",
       "      <td>ATGCCCCAGCTGTGCACCAGGGCCAGCTCATCAGAGATTCCTGGCA...</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CCDS746.1|Hs109|chr1</td>\n",
       "      <td>GCLM</td>\n",
       "      <td>TACATTTTACAAGCTAAAAGAAGGGGT</td>\n",
       "      <td>ATGGGCACCGACAGCCGCGCGGCCAAGGCGCTCCTGGCGCGGGCCC...</td>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>CCDS87159.1|Hs109|chr3</td>\n",
       "      <td>SLC33A1</td>\n",
       "      <td>AATTTAAAAAGTTACAGGATGAAGGAT</td>\n",
       "      <td>ATGTCACCCACCATCTCCCACAAGGACAGCAGCCGGCAACGGCGGC...</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>CCDS87280.1|Hs109|chr4</td>\n",
       "      <td>HPGD</td>\n",
       "      <td>CCCAGGGTTACTTTAGAACTGCTGGAT</td>\n",
       "      <td>ATGCACGTGAACGGCAAAGTGGCGCTGGTGACCGGCGCGGCTCAGG...</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>CCDS87397.1|Hs109|chr6</td>\n",
       "      <td>OARD1</td>\n",
       "      <td>GCGATATATATATTACTTGGATTGGAT</td>\n",
       "      <td>ATGGCCAGCAGCCTTAATGAAGATCCAGAAGGAAGCAGAATCACTT...</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>CCDS87482.1|Hs109|chr7</td>\n",
       "      <td>BZW2</td>\n",
       "      <td>AGTGGTTACAAAATGCAGAAGAAGAAT</td>\n",
       "      <td>ATGGCAGAAAAAGATGCCAACTCTGTTACCTCGTCTTTGAGAAAAG...</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>CCDS87676.1|Hs109|chr9</td>\n",
       "      <td>INIP</td>\n",
       "      <td>CTGTTTTACCTCGCCTTGACCCAGAAT</td>\n",
       "      <td>ATGCAGAACCAGTCTTCAACAAATCATCCTGGAGCTAGCATTGCAC...</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>354 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  CCDS Name Gene Name                         gRNA  \\\n",
       "0      CCDS157.1|Hs109|chr1    CELA2A  GCGGGTCTCCAATTACATCGACTGGAT   \n",
       "1      CCDS450.1|Hs109|chr1    COL9A2  CCTCTGCCCGCCTTACAGAGCCTGGAT   \n",
       "2      CCDS627.1|Hs109|chr1     UBE2U  CAGTATTACAAATGGAAGAAAATGGAT   \n",
       "3      CCDS694.1|Hs109|chr1   DNASE2B  GGCAAATTTACCAAGCATTTCAAGGAT   \n",
       "4      CCDS746.1|Hs109|chr1      GCLM  TACATTTTACAAGCTAAAAGAAGGGGT   \n",
       "..                      ...       ...                          ...   \n",
       "349  CCDS87159.1|Hs109|chr3   SLC33A1  AATTTAAAAAGTTACAGGATGAAGGAT   \n",
       "350  CCDS87280.1|Hs109|chr4      HPGD  CCCAGGGTTACTTTAGAACTGCTGGAT   \n",
       "351  CCDS87397.1|Hs109|chr6     OARD1  GCGATATATATATTACTTGGATTGGAT   \n",
       "352  CCDS87482.1|Hs109|chr7      BZW2  AGTGGTTACAAAATGCAGAAGAAGAAT   \n",
       "353  CCDS87676.1|Hs109|chr9      INIP  CTGTTTTACCTCGCCTTGACCCAGAAT   \n",
       "\n",
       "                                              Sequence  TTAC position  Frame  \\\n",
       "0    ATGATAAGGACGCTGCTGCTGTCCACTTTGGTGGCTGGAGCCCTCA...              9      1   \n",
       "1    ATGGCCGCCGCTACGGCCTCCCCCCGCAGCCTCCTTGTTCTCCTCC...              9      3   \n",
       "2    ATGCACGGCAGAGCTTACCTCTTGCTGCACAGAGACTTCTGTGATC...             16      1   \n",
       "3    ATGCCCCAGCTGTGCACCAGGGCCAGCTCATCAGAGATTCCTGGCA...             14      1   \n",
       "4    ATGGGCACCGACAGCCGCGCGGCCAAGGCGCTCCTGGCGCGGGCCC...             15      2   \n",
       "..                                                 ...            ...    ...   \n",
       "349  ATGTCACCCACCATCTCCCACAAGGACAGCAGCCGGCAACGGCGGC...             10      2   \n",
       "350  ATGCACGTGAACGGCAAAGTGGCGCTGGTGACCGGCGCGGCTCAGG...             14      1   \n",
       "351  ATGGCCAGCAGCCTTAATGAAGATCCAGAAGGAAGCAGAATCACTT...              9      1   \n",
       "352  ATGGCAGAAAAAGATGCCAACTCTGTTACCTCGTCTTTGAGAAAAG...             16      2   \n",
       "353  ATGCAGAACCAGTCTTCAACAAATCATCCTGGAGCTAGCATTGCAC...             16      2   \n",
       "\n",
       "     Distance from 3' End  \n",
       "0                      36  \n",
       "1                      28  \n",
       "2                      51  \n",
       "3                      45  \n",
       "4                      23  \n",
       "..                    ...  \n",
       "349                    44  \n",
       "350                    30  \n",
       "351                    21  \n",
       "352                    44  \n",
       "353                    20  \n",
       "\n",
       "[354 rows x 7 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#SaCas9 search in the forward direction, TTACs in window of 8-17, any frame\n",
    "\n",
    "# Parameters:\n",
    "search_regex = \"TTAC\\w{6,15}G[G|A][G|A]T\" #Regex for sequence to match\n",
    "distance_from_end = 60 #Maximum distance from 3' end of gene to TTAC = last 20 aas\n",
    "gRNA_length = 27 #Includes PAM\n",
    "PAM_length = 6\n",
    "file_name = \"SaCas9_8-17_Forward_20aa\"\n",
    "\n",
    "'''\n",
    "SaCas9 has a gRNA structure of GN(21)NNGRRT\n",
    "PAM regex is thus \"\\w{2}G(G|A)(G|A)T\"\n",
    "Want TTAC from window of 8-17 -> \"TTAC\\w{6-15}G[G|A][G|A]T\"\n",
    "'''\n",
    "\n",
    "#Initialize relevant variables\n",
    "counter = 0\n",
    "good_sequences = []\n",
    "filtered_data = []\n",
    "\n",
    "#Load file\n",
    "genesequences = SeqIO.parse(\"CCDS_nucleotide.current.fna\", \"fasta\")\n",
    "pbar = tqdm(genesequences)\n",
    "\n",
    "#Iterate through sequences and search for matches\n",
    "for sequence in pbar: \n",
    "    target_RC_search = re.finditer(search_regex, str(sequence.seq)) #Regex match iterator\n",
    "    for match in target_RC_search:\n",
    "        (start, stop) = match.span()\n",
    "        real_start = start-(gRNA_length-len(match.group())) #Compensates for parts of gRNA missing from match\n",
    "        if real_start+1 >= len(sequence)-distance_from_end:\n",
    "            TTAC_pos = len(match.group())-PAM_length\n",
    "            frame = (start+4)%3 + 1\n",
    "            gRNA_dist = len(sequence)-start-4\n",
    "            gRNA_sequence = str(sequence.seq)[real_start:real_start+gRNA_length]\n",
    "            gene_name = gene_dict[re.match(\"CCDS.*\\.\\d\", sequence.name).group()]\n",
    "            good_sequences.append(sequence)\n",
    "            filtered_data.append([sequence.name, gene_name, gRNA_sequence, str(sequence.seq), TTAC_pos, frame, gRNA_dist])\n",
    "            counter += 1\n",
    "    \n",
    "print(f\"Number of matching sequences: {counter}\")\n",
    "\n",
    "#Write data to relevant files\n",
    "with open(file_name+\"_Matches.fasta\", \"w\") as output_handle:\n",
    "        SeqIO.write(good_sequences, output_handle, \"fasta\")\n",
    "        \n",
    "df = pd.DataFrame(filtered_data, columns = [\"CCDS Name\", \"Gene Name\", \"gRNA\", \"Sequence\", \"TTAC position\", \"Frame\", \"Distance from 3' End\"])\n",
    "df.to_csv(file_name+\".csv\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6e3264dc36b34d5e885b934556010da1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=1.0, bar_style='info', max=1.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of matching sequences: 90\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CCDS Name</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>gRNA</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>TTAC position</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Distance from 3' End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCDS157.1|Hs109|chr1</td>\n",
       "      <td>CELA2A</td>\n",
       "      <td>TTACATCGACTGGATCAATTCGG</td>\n",
       "      <td>ATGATAAGGACGCTGCTGCTGTCCACTTTGGTGGCTGGAGCCCTCA...</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCDS348.1|Hs109|chr1</td>\n",
       "      <td>PTP4A2</td>\n",
       "      <td>TTACGCTTCAGAGATACCAATGG</td>\n",
       "      <td>ATGAACCGTCCAGCCCCTGTGGAGATCTCCTATGAGAACATGCGTT...</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCDS442.1|Hs109|chr1</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>TTACCAGCCAGCCGAGGTCCTGG</td>\n",
       "      <td>ATGGCCACCACCAAGCGCGTCTTGTACGTGGGTGGACTGGCAGAGG...</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CCDS1059.1|Hs109|chr1</td>\n",
       "      <td>RPS27</td>\n",
       "      <td>TTACAGAAGGATGTTCCTTCAGG</td>\n",
       "      <td>ATGCCTCTCGCAAAGGATCTCCTTCATCCCTCTCCAGAAGAGGAGA...</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CCDS1230.1|Hs109|chr1</td>\n",
       "      <td>SDHC</td>\n",
       "      <td>TTACTGTGTTGTCCTCTATGGGG</td>\n",
       "      <td>ATGGCTGCGCTGTTGCTGAGACACGTTGGTCGTCATTGCCTCCGAG...</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>CCDS76767.1|Hs109|chr15</td>\n",
       "      <td>CA12</td>\n",
       "      <td>TTACAAGCCAGCCACCAAGATGG</td>\n",
       "      <td>ATGCCCCGGCGCAGCCTGCACGCGGCGGCCGTGCTCCTGCTGGTGA...</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>CCDS77519.1|Hs109|chr2</td>\n",
       "      <td>KANSL1L</td>\n",
       "      <td>TTACTAATGTTAAAAAAAATAGG</td>\n",
       "      <td>ATGACCCCAGCTCTGAGGGAGGCAACAGCAAAGGGTATCAGCTTTT...</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>CCDS78135.1|Hs109|chr6</td>\n",
       "      <td>SAYSD1</td>\n",
       "      <td>TTACAGTTGAGACCCCTGGCAGG</td>\n",
       "      <td>ATGCTGGAAGCGGCTCAGCCCCAGGGCAGCACATCAGAGACACCAT...</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>CCDS82507.1|Hs109|chr2</td>\n",
       "      <td>TEX51</td>\n",
       "      <td>TTACTGGCAAAGGAAGAAGGAGG</td>\n",
       "      <td>ATGCTGCCTCTCCTGATCATCTGTCTCCTGCCTGCCATTGAAGGGA...</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>CCDS86446.1|Hs109|chr15</td>\n",
       "      <td>C15orf41</td>\n",
       "      <td>TTACAAGTACCAGTTGATTTGGG</td>\n",
       "      <td>ATGATACTGACCAAAGCTCAGTACGACGAGATAGCCCAGTGCCTAG...</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  CCDS Name Gene Name                     gRNA  \\\n",
       "0      CCDS157.1|Hs109|chr1    CELA2A  TTACATCGACTGGATCAATTCGG   \n",
       "1      CCDS348.1|Hs109|chr1    PTP4A2  TTACGCTTCAGAGATACCAATGG   \n",
       "2      CCDS442.1|Hs109|chr1      PPIE  TTACCAGCCAGCCGAGGTCCTGG   \n",
       "3     CCDS1059.1|Hs109|chr1     RPS27  TTACAGAAGGATGTTCCTTCAGG   \n",
       "4     CCDS1230.1|Hs109|chr1      SDHC  TTACTGTGTTGTCCTCTATGGGG   \n",
       "..                      ...       ...                      ...   \n",
       "85  CCDS76767.1|Hs109|chr15      CA12  TTACAAGCCAGCCACCAAGATGG   \n",
       "86   CCDS77519.1|Hs109|chr2   KANSL1L  TTACTAATGTTAAAAAAAATAGG   \n",
       "87   CCDS78135.1|Hs109|chr6    SAYSD1  TTACAGTTGAGACCCCTGGCAGG   \n",
       "88   CCDS82507.1|Hs109|chr2     TEX51  TTACTGGCAAAGGAAGAAGGAGG   \n",
       "89  CCDS86446.1|Hs109|chr15  C15orf41  TTACAAGTACCAGTTGATTTGGG   \n",
       "\n",
       "                                             Sequence  TTAC position  Frame  \\\n",
       "0   ATGATAAGGACGCTGCTGCTGTCCACTTTGGTGGCTGGAGCCCTCA...             20      1   \n",
       "1   ATGAACCGTCCAGCCCCTGTGGAGATCTCCTATGAGAACATGCGTT...             20      2   \n",
       "2   ATGGCCACCACCAAGCGCGTCTTGTACGTGGGTGGACTGGCAGAGG...             20      3   \n",
       "3   ATGCCTCTCGCAAAGGATCTCCTTCATCCCTCTCCAGAAGAGGAGA...             20      3   \n",
       "4   ATGGCTGCGCTGTTGCTGAGACACGTTGGTCGTCATTGCCTCCGAG...             20      3   \n",
       "..                                                ...            ...    ...   \n",
       "85  ATGCCCCGGCGCAGCCTGCACGCGGCGGCCGTGCTCCTGCTGGTGA...             20      1   \n",
       "86  ATGACCCCAGCTCTGAGGGAGGCAACAGCAAAGGGTATCAGCTTTT...             20      3   \n",
       "87  ATGCTGGAAGCGGCTCAGCCCCAGGGCAGCACATCAGAGACACCAT...             20      2   \n",
       "88  ATGCTGCCTCTCCTGATCATCTGTCTCCTGCCTGCCATTGAAGGGA...             20      3   \n",
       "89  ATGATACTGACCAAAGCTCAGTACGACGAGATAGCCCAGTGCCTAG...             20      2   \n",
       "\n",
       "    Distance from 3' End  \n",
       "0                     36  \n",
       "1                     38  \n",
       "2                     31  \n",
       "3                     34  \n",
       "4                     34  \n",
       "..                   ...  \n",
       "85                    39  \n",
       "86                    22  \n",
       "87                    26  \n",
       "88                    25  \n",
       "89                    29  \n",
       "\n",
       "[90 rows x 7 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#SpCas9 search in the forward direction, TTAC at position 20, any frame\n",
    "\n",
    "# Parameters:\n",
    "search_regex = \"TTAC\\w{17}GG\" #Regex for sequence to match\n",
    "distance_from_end = 45 #Maximum distance from 3' end of gene to TTAC = last 15 aas\n",
    "gRNA_length = 23 #Includes PAM\n",
    "PAM_length = 3\n",
    "file_name = \"SpCas9_20_Forward_15aa\"\n",
    "\n",
    "#Initialize relevant variables\n",
    "counter = 0\n",
    "good_sequences = []\n",
    "filtered_data = []\n",
    "\n",
    "#Load file\n",
    "genesequences = SeqIO.parse(\"CCDS_nucleotide.current.fna\", \"fasta\")\n",
    "pbar = tqdm(genesequences)\n",
    "\n",
    "#Iterate through sequences and search for matches\n",
    "for sequence in pbar: \n",
    "    target_RC_search = re.finditer(search_regex, str(sequence.seq)) #Regex match iterator\n",
    "    for match in target_RC_search:\n",
    "        (start, stop) = match.span()\n",
    "        real_start = start-(gRNA_length-len(match.group())) #Compensates for parts of gRNA missing from match\n",
    "        if real_start+1 >= len(sequence)-distance_from_end:\n",
    "            TTAC_pos = len(match.group())-PAM_length\n",
    "            frame = (start+4)%3 + 1\n",
    "            gRNA_dist = len(sequence)-start-4\n",
    "            gRNA_sequence = str(sequence.seq)[real_start:real_start+gRNA_length]\n",
    "            gene_name = gene_dict[re.match(\"CCDS.*\\.\\d\", sequence.name).group()]\n",
    "            good_sequences.append(sequence)\n",
    "            filtered_data.append([sequence.name, gene_name, gRNA_sequence, str(sequence.seq), TTAC_pos, frame, gRNA_dist])\n",
    "            counter += 1\n",
    "    \n",
    "print(f\"Number of matching sequences: {counter}\")\n",
    "\n",
    "#Write data to relevant files\n",
    "with open(file_name+\"_Matches.fasta\", \"w\") as output_handle:\n",
    "        SeqIO.write(good_sequences, output_handle, \"fasta\")\n",
    "        \n",
    "df = pd.DataFrame(filtered_data, columns = [\"CCDS Name\", \"Gene Name\", \"gRNA\", \"Sequence\", \"TTAC position\", \"Frame\", \"Distance from 3' End\"])\n",
    "df.to_csv(file_name+\".csv\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "64d6a644331c4d3995c37a6fb2ee3694",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=1.0, bar_style='info', max=1.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of matching sequences: 136\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CCDS Name</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>gRNA</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>TTAC position</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Distance from 3' End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCDS157.1|Hs109|chr1</td>\n",
       "      <td>CELA2A</td>\n",
       "      <td>TTACATCGACTGGATCAATTCGG</td>\n",
       "      <td>ATGATAAGGACGCTGCTGCTGTCCACTTTGGTGGCTGGAGCCCTCA...</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCDS348.1|Hs109|chr1</td>\n",
       "      <td>PTP4A2</td>\n",
       "      <td>TTACGCTTCAGAGATACCAATGG</td>\n",
       "      <td>ATGAACCGTCCAGCCCCTGTGGAGATCTCCTATGAGAACATGCGTT...</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCDS442.1|Hs109|chr1</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>TTACCAGCCAGCCGAGGTCCTGG</td>\n",
       "      <td>ATGGCCACCACCAAGCGCGTCTTGTACGTGGGTGGACTGGCAGAGG...</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CCDS1059.1|Hs109|chr1</td>\n",
       "      <td>RPS27</td>\n",
       "      <td>TTACAGAAGGATGTTCCTTCAGG</td>\n",
       "      <td>ATGCCTCTCGCAAAGGATCTCCTTCATCCCTCTCCAGAAGAGGAGA...</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CCDS1230.1|Hs109|chr1</td>\n",
       "      <td>SDHC</td>\n",
       "      <td>TTACTGTGTTGTCCTCTATGGGG</td>\n",
       "      <td>ATGGCTGCGCTGTTGCTGAGACACGTTGGTCGTCATTGCCTCCGAG...</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>CCDS77519.1|Hs109|chr2</td>\n",
       "      <td>KANSL1L</td>\n",
       "      <td>TTACTAATGTTAAAAAAAATAGG</td>\n",
       "      <td>ATGACCCCAGCTCTGAGGGAGGCAACAGCAAAGGGTATCAGCTTTT...</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>CCDS78135.1|Hs109|chr6</td>\n",
       "      <td>SAYSD1</td>\n",
       "      <td>TTACAGTTGAGACCCCTGGCAGG</td>\n",
       "      <td>ATGCTGGAAGCGGCTCAGCCCCAGGGCAGCACATCAGAGACACCAT...</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>CCDS82180.1|Hs109|chr17</td>\n",
       "      <td>TBX4</td>\n",
       "      <td>TTACAGTACCATTCAGGAATGGG</td>\n",
       "      <td>ATGCTGCAGGATAAGGGCCTGTCCGAGAGCGAGGAGGCCTTCCGGG...</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>CCDS82507.1|Hs109|chr2</td>\n",
       "      <td>TEX51</td>\n",
       "      <td>TTACTGGCAAAGGAAGAAGGAGG</td>\n",
       "      <td>ATGCTGCCTCTCCTGATCATCTGTCTCCTGCCTGCCATTGAAGGGA...</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>CCDS86446.1|Hs109|chr15</td>\n",
       "      <td>C15orf41</td>\n",
       "      <td>TTACAAGTACCAGTTGATTTGGG</td>\n",
       "      <td>ATGATACTGACCAAAGCTCAGTACGACGAGATAGCCCAGTGCCTAG...</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>136 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   CCDS Name Gene Name                     gRNA  \\\n",
       "0       CCDS157.1|Hs109|chr1    CELA2A  TTACATCGACTGGATCAATTCGG   \n",
       "1       CCDS348.1|Hs109|chr1    PTP4A2  TTACGCTTCAGAGATACCAATGG   \n",
       "2       CCDS442.1|Hs109|chr1      PPIE  TTACCAGCCAGCCGAGGTCCTGG   \n",
       "3      CCDS1059.1|Hs109|chr1     RPS27  TTACAGAAGGATGTTCCTTCAGG   \n",
       "4      CCDS1230.1|Hs109|chr1      SDHC  TTACTGTGTTGTCCTCTATGGGG   \n",
       "..                       ...       ...                      ...   \n",
       "131   CCDS77519.1|Hs109|chr2   KANSL1L  TTACTAATGTTAAAAAAAATAGG   \n",
       "132   CCDS78135.1|Hs109|chr6    SAYSD1  TTACAGTTGAGACCCCTGGCAGG   \n",
       "133  CCDS82180.1|Hs109|chr17      TBX4  TTACAGTACCATTCAGGAATGGG   \n",
       "134   CCDS82507.1|Hs109|chr2     TEX51  TTACTGGCAAAGGAAGAAGGAGG   \n",
       "135  CCDS86446.1|Hs109|chr15  C15orf41  TTACAAGTACCAGTTGATTTGGG   \n",
       "\n",
       "                                              Sequence  TTAC position  Frame  \\\n",
       "0    ATGATAAGGACGCTGCTGCTGTCCACTTTGGTGGCTGGAGCCCTCA...             20      1   \n",
       "1    ATGAACCGTCCAGCCCCTGTGGAGATCTCCTATGAGAACATGCGTT...             20      2   \n",
       "2    ATGGCCACCACCAAGCGCGTCTTGTACGTGGGTGGACTGGCAGAGG...             20      3   \n",
       "3    ATGCCTCTCGCAAAGGATCTCCTTCATCCCTCTCCAGAAGAGGAGA...             20      3   \n",
       "4    ATGGCTGCGCTGTTGCTGAGACACGTTGGTCGTCATTGCCTCCGAG...             20      3   \n",
       "..                                                 ...            ...    ...   \n",
       "131  ATGACCCCAGCTCTGAGGGAGGCAACAGCAAAGGGTATCAGCTTTT...             20      3   \n",
       "132  ATGCTGGAAGCGGCTCAGCCCCAGGGCAGCACATCAGAGACACCAT...             20      2   \n",
       "133  ATGCTGCAGGATAAGGGCCTGTCCGAGAGCGAGGAGGCCTTCCGGG...             20      2   \n",
       "134  ATGCTGCCTCTCCTGATCATCTGTCTCCTGCCTGCCATTGAAGGGA...             20      3   \n",
       "135  ATGATACTGACCAAAGCTCAGTACGACGAGATAGCCCAGTGCCTAG...             20      2   \n",
       "\n",
       "     Distance from 3' End  \n",
       "0                      36  \n",
       "1                      38  \n",
       "2                      31  \n",
       "3                      34  \n",
       "4                      34  \n",
       "..                    ...  \n",
       "131                    22  \n",
       "132                    26  \n",
       "133                    47  \n",
       "134                    25  \n",
       "135                    29  \n",
       "\n",
       "[136 rows x 7 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#SpCas9 search in the forward direction, TTAC at position 20, any frame\n",
    "\n",
    "#Parameters:\n",
    "search_regex = \"TTAC\\w{17}GG\" #Regex for sequence to match\n",
    "distance_from_end = 60 #Maximum distance from 3' end of gene to TTAC = last 20 aas\n",
    "gRNA_length = 23 #Includes PAM\n",
    "PAM_length = 3\n",
    "file_name = \"SpCas9_20_Forward_20aa\"\n",
    "\n",
    "#Initialize relevant variables\n",
    "counter = 0\n",
    "good_sequences = []\n",
    "filtered_data = []\n",
    "\n",
    "#Load file\n",
    "genesequences = SeqIO.parse(\"CCDS_nucleotide.current.fna\", \"fasta\")\n",
    "pbar = tqdm(genesequences)\n",
    "\n",
    "#Iterate through sequences and search for matches\n",
    "for sequence in pbar: \n",
    "    target_RC_search = re.finditer(search_regex, str(sequence.seq)) #Regex match iterator\n",
    "    for match in target_RC_search:\n",
    "        (start, stop) = match.span()\n",
    "        real_start = start-(gRNA_length-len(match.group())) #Compensates for parts of gRNA missing from match\n",
    "        if real_start+1 >= len(sequence)-distance_from_end:\n",
    "            TTAC_pos = len(match.group())-PAM_length\n",
    "            frame = (start+4)%3 + 1\n",
    "            gRNA_dist = len(sequence)-start-4\n",
    "            gRNA_sequence = str(sequence.seq)[real_start:real_start+gRNA_length]\n",
    "            gene_name = gene_dict[re.match(\"CCDS.*\\.\\d\", sequence.name).group()]\n",
    "            good_sequences.append(sequence)\n",
    "            filtered_data.append([sequence.name, gene_name, gRNA_sequence, str(sequence.seq), TTAC_pos, frame, gRNA_dist])\n",
    "            counter += 1\n",
    "    \n",
    "print(f\"Number of matching sequences: {counter}\")\n",
    "\n",
    "#Write data to relevant files\n",
    "with open(file_name+\"_Matches.fasta\", \"w\") as output_handle:\n",
    "        SeqIO.write(good_sequences, output_handle, \"fasta\")\n",
    "        \n",
    "df = pd.DataFrame(filtered_data, columns = [\"CCDS Name\", \"Gene Name\", \"gRNA\", \"Sequence\", \"TTAC position\", \"Frame\", \"Distance from 3' End\"])\n",
    "df.to_csv(file_name+\".csv\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7dc5e72730524fcbadfb58fa91e4577a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=1.0, bar_style='info', max=1.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of matching sequences: 1146\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CCDS Name</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>gRNA</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Distance from 3' End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCDS246.1|Hs109|chr1</td>\n",
       "      <td>PNRC2</td>\n",
       "      <td>TTACTTAAAGTACAGGTATAA</td>\n",
       "      <td>ATGGGTGGTGGAGAGAGGTATAACATTCCAGCCCCTCAATCTAGAA...</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCDS248.1|Hs109|chr1</td>\n",
       "      <td>IFNLR1</td>\n",
       "      <td>TTACATGGCCAGGTGA</td>\n",
       "      <td>ATGGCGGGGCCCGAGCGCTGGGGCCCCCTGCTCCTGTGCCTGCTGC...</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCDS249.1|Hs109|chr1</td>\n",
       "      <td>IFNLR1</td>\n",
       "      <td>TTACATGGCCAGGTGA</td>\n",
       "      <td>ATGGCGGGGCCCGAGCGCTGGGGCCCCCTGCTCCTGTGCCTGCTGC...</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CCDS406.1|Hs109|chr1</td>\n",
       "      <td>EVA1B</td>\n",
       "      <td>TTACTGA</td>\n",
       "      <td>ATGGATGCCCCGCGAAGGGACATGGAGTTGCTCAGCAACAGCCTGG...</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CCDS467.1|Hs109|chr1</td>\n",
       "      <td>ZMYND12</td>\n",
       "      <td>TTACTTAG</td>\n",
       "      <td>ATGAATGTGATCTACCCACTGGCAGTCCCCAAGGGGCGCAGACTCT...</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1141</th>\n",
       "      <td>CCDS87332.1|Hs109|chr5</td>\n",
       "      <td>ARHGAP26</td>\n",
       "      <td>TTACGTGGAGTTCCTCTAA</td>\n",
       "      <td>ATGGAGCGTGCCCTGCCGAGGGGGCGCTGCCTCCCCTTGGGAAAAG...</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1142</th>\n",
       "      <td>CCDS87340.1|Hs109|chr5</td>\n",
       "      <td>PPP1R2B</td>\n",
       "      <td>TTACGAAGTTCATAG</td>\n",
       "      <td>ATGGCGGCCTCGACGGCCTCCCACCGGCCCATCAAGGGGATCTTGA...</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1143</th>\n",
       "      <td>CCDS87395.1|Hs109|chr6</td>\n",
       "      <td>OARD1</td>\n",
       "      <td>TTACTGTGTACACACTCTGA</td>\n",
       "      <td>ATGGGCGCTGGGATAGCTGTCCTCTTTAAGAAGAAATTTGGAGGGG...</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1144</th>\n",
       "      <td>CCDS87430.1|Hs109|chr6</td>\n",
       "      <td>CD164</td>\n",
       "      <td>TTACCACACTCTGTAA</td>\n",
       "      <td>ATGCGAAAAGGAAGAAAAGTACCCATGTATGTACCGGGTGTTTTGC...</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1145</th>\n",
       "      <td>CCDS87484.1|Hs109|chr7</td>\n",
       "      <td>TMEM196</td>\n",
       "      <td>TTACCTACAAGGAAATAA</td>\n",
       "      <td>ATGTGCACCAGCGGTCAGATTATTGGGAGCCTCTTGGTGCTCTCCG...</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1146 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   CCDS Name Gene Name                   gRNA  \\\n",
       "0       CCDS246.1|Hs109|chr1     PNRC2  TTACTTAAAGTACAGGTATAA   \n",
       "1       CCDS248.1|Hs109|chr1    IFNLR1       TTACATGGCCAGGTGA   \n",
       "2       CCDS249.1|Hs109|chr1    IFNLR1       TTACATGGCCAGGTGA   \n",
       "3       CCDS406.1|Hs109|chr1     EVA1B                TTACTGA   \n",
       "4       CCDS467.1|Hs109|chr1   ZMYND12               TTACTTAG   \n",
       "...                      ...       ...                    ...   \n",
       "1141  CCDS87332.1|Hs109|chr5  ARHGAP26    TTACGTGGAGTTCCTCTAA   \n",
       "1142  CCDS87340.1|Hs109|chr5   PPP1R2B        TTACGAAGTTCATAG   \n",
       "1143  CCDS87395.1|Hs109|chr6     OARD1   TTACTGTGTACACACTCTGA   \n",
       "1144  CCDS87430.1|Hs109|chr6     CD164       TTACCACACTCTGTAA   \n",
       "1145  CCDS87484.1|Hs109|chr7   TMEM196     TTACCTACAAGGAAATAA   \n",
       "\n",
       "                                               Sequence  Frame  \\\n",
       "0     ATGGGTGGTGGAGAGAGGTATAACATTCCAGCCCCTCAATCTAGAA...      2   \n",
       "1     ATGGCGGGGCCCGAGCGCTGGGGCCCCCTGCTCCTGTGCCTGCTGC...      1   \n",
       "2     ATGGCGGGGCCCGAGCGCTGGGGCCCCCTGCTCCTGTGCCTGCTGC...      1   \n",
       "3     ATGGATGCCCCGCGAAGGGACATGGAGTTGCTCAGCAACAGCCTGG...      1   \n",
       "4     ATGAATGTGATCTACCCACTGGCAGTCCCCAAGGGGCGCAGACTCT...      3   \n",
       "...                                                 ...    ...   \n",
       "1141  ATGGAGCGTGCCCTGCCGAGGGGGCGCTGCCTCCCCTTGGGAAAAG...      1   \n",
       "1142  ATGGCGGCCTCGACGGCCTCCCACCGGCCCATCAAGGGGATCTTGA...      2   \n",
       "1143  ATGGGCGCTGGGATAGCTGTCCTCTTTAAGAAGAAATTTGGAGGGG...      3   \n",
       "1144  ATGCGAAAAGGAAGAAAAGTACCCATGTATGTACCGGGTGTTTTGC...      1   \n",
       "1145  ATGTGCACCAGCGGTCAGATTATTGGGAGCCTCTTGGTGCTCTCCG...      2   \n",
       "\n",
       "      Distance from 3' End  \n",
       "0                       17  \n",
       "1                       12  \n",
       "2                       12  \n",
       "3                        3  \n",
       "4                        4  \n",
       "...                    ...  \n",
       "1141                    15  \n",
       "1142                    11  \n",
       "1143                    16  \n",
       "1144                    12  \n",
       "1145                    14  \n",
       "\n",
       "[1146 rows x 6 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Search in the forward direction for SpCas9 gRNAs that overlap with stop/5' UTR, TTAC at position 20, any frame\n",
    "\n",
    "#Parameters:\n",
    "search_regex = \"(TTAC\\w{0,17}$)|(TTAC\\w{17}G$)\" #Regex for sequence to match\n",
    "file_name = \"5_UTR_Overlapping_Forward_SpCas9\"\n",
    "\n",
    "#Initialize relevant variables\n",
    "counter = 0\n",
    "good_sequences = []\n",
    "filtered_data = []\n",
    "\n",
    "#Load file\n",
    "genesequences = SeqIO.parse(\"CCDS_nucleotide.current.fna\", \"fasta\")\n",
    "pbar = tqdm(genesequences)\n",
    "\n",
    "#Iterate through sequences and search for matches\n",
    "for sequence in pbar: \n",
    "    last24 = str(sequence.seq)[-24:] #If you don't truncate, regex search totally kills your memory :(\n",
    "    target_RC_search = re.finditer(search_regex, last24) #Regex match iterator\n",
    "    for match in target_RC_search:\n",
    "        (start, stop) = match.span()\n",
    "        frame = (start+4)%3 + 1\n",
    "        gRNA_dist = 24-start-4\n",
    "        gRNA_sequence = last24[start:]\n",
    "        gene_name = gene_dict[re.match(\"CCDS.*\\.\\d\", sequence.name).group()]\n",
    "        good_sequences.append(sequence)\n",
    "        filtered_data.append([sequence.name, gene_name, gRNA_sequence, str(sequence.seq), frame, gRNA_dist])\n",
    "        counter += 1\n",
    "    \n",
    "print(f\"Number of matching sequences: {counter}\")\n",
    "\n",
    "#Write data to relevant files\n",
    "with open(file_name+\"_Matches.fasta\", \"w\") as output_handle:\n",
    "        SeqIO.write(good_sequences, output_handle, \"fasta\")\n",
    "\n",
    "initial_filtered_sequences = good_sequences[:]\n",
    "\n",
    "df = pd.DataFrame(filtered_data, columns = [\"CCDS Name\", \"Gene Name\", \"gRNA\", \"Sequence\", \"Frame\", \"Distance from 3' End\"])\n",
    "df.to_csv(file_name+\".csv\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: CCDS246.1|Hs109|chr1\n",
      "Name: CCDS246.1|Hs109|chr1\n",
      "Description: CCDS246.1|Hs109|chr1\n",
      "Number of features: 0\n",
      "Seq('ATGGGTGGTGGAGAGAGGTATAACATTCCAGCCCCTCAATCTAGAAATGTTAGT...TAA', SingleLetterAlphabet())\n",
      "CCDS246\n"
     ]
    }
   ],
   "source": [
    "print(initial_filtered_sequences[0])\n",
    "current_CCDS_id = re.match(\"(CCDS.*)\\.\\d\", str(initial_filtered_sequences[0].name)).group(1)\n",
    "print(current_CCDS_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "74f9edfe99894bb4a3327a8ef0df960c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=1.0, bar_style='info', max=1.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "996\n"
     ]
    }
   ],
   "source": [
    "#IGNORE THIS - IS DEPRECATED, kept for comments\n",
    "'''\n",
    "#Create dictionary of CDS to ENST, and a filtered set of ENST sequence names\n",
    "good_ENSTs = set()\n",
    "CCDStoENST = {}\n",
    "\n",
    "with open('5UTR_CCDS_to_ENST.csv', newline='') as file:\n",
    "    linereader = csv.reader(file, delimiter=' ', quotechar='|')\n",
    "    for line in linereader:\n",
    "        CCDS_id, ENST_id = line[0].split(',')\n",
    "        good_ENSTs.add(ENST_id)\n",
    "        CCDStoENST[CCDS_id]=ENST_id\n",
    "        \n",
    "#Filter ENST sequences for faster parsing:\n",
    "good_ENST_sequences = {}\n",
    "\n",
    "#Load file\n",
    "ENSTsequences = SeqIO.parse(\"gencode.v38.pc_transcripts.fa\", \"fasta\")\n",
    "pbar = tqdm(ENSTsequences)\n",
    "\n",
    "#Iterate through sequences and search for matches\n",
    "for sequence in pbar: \n",
    "    ENSTname = re.match(\"ENST\\d*\\.\\d\",sequence.name).group()\n",
    "    if ENSTname in good_ENSTs:\n",
    "        good_ENST_sequences[ENSTname] = sequence\n",
    "        \n",
    "print(len(good_ENST_sequences))\n",
    "'''\n",
    "#Somehow there are sequences lost here, went from 1129 CDS to ENST matches to 996 filtered ENST sequences\n",
    "#I suspect this is a version issue, so will instead try with no version numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "57d195a3eb1c4c92b7a0dd15b2e46f72",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=1.0, bar_style='info', max=1.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "1126\n"
     ]
    }
   ],
   "source": [
    "#Create dictionary of CDS to ENST, and a filtered set of ENST sequence names\n",
    "good_ENSTs = set()\n",
    "CCDStoENST = {}\n",
    "\n",
    "with open('5UTR_CCDS_to_ENST.csv', newline='') as file:\n",
    "    linereader = csv.reader(file, delimiter=' ', quotechar='|')\n",
    "    for line in linereader:\n",
    "        CCDS_id, ENST_id_initial = line[0].split(',')\n",
    "        ENST_id = re.match(\"(ENST\\d*)\\.\\d\", ENST_id_initial).group(1)\n",
    "        good_ENSTs.add(ENST_id)\n",
    "        CCDStoENST[CCDS_id]=ENST_id\n",
    "        \n",
    "#Filter ENST sequences for faster parsing:\n",
    "good_ENST_sequences = {}\n",
    "\n",
    "#Load file\n",
    "ENSTsequences = SeqIO.parse(\"gencode.v38.pc_transcripts.fa\", \"fasta\")\n",
    "pbar = tqdm(ENSTsequences)\n",
    "\n",
    "#Iterate through sequences and search for matches\n",
    "for sequence in pbar: \n",
    "    ENSTname = re.match(\"(ENST\\d*)\\.\\d\",sequence.name).group(1)\n",
    "    if ENSTname in good_ENSTs:\n",
    "        good_ENST_sequences[ENSTname] = sequence\n",
    "        \n",
    "print(len(good_ENST_sequences))\n",
    "#Now, we are up to 1126 out of 1129. That is good enough for me. \n",
    "#One missing is from key issues reading in the start of .csv for CCDS\n",
    "#Not sure about the other two"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "633aa2f203524a4f80e55f5641e0c8fc",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=1146.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of matching sequences: 76\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CCDS Name</th>\n",
       "      <th>ENST Name</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>gRNA</th>\n",
       "      <th>CCDS Sequence</th>\n",
       "      <th>ENST Sequence</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Distance from 3' End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCDS503.1|Hs109|chr1</td>\n",
       "      <td>ENST00000372343</td>\n",
       "      <td>IPO13</td>\n",
       "      <td>TTACACAGCTGACTACTGAGGGG</td>\n",
       "      <td>ATGGAGCGGCGGGAGGAGCAGCCGGGGGCTGCAGGGGCTGGAGCAG...</td>\n",
       "      <td>GCTTGTCTTGTCAGTCACTGGGGCGGAGGCAGCGGCTGTAGCGGGG...</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCDS722.1|Hs109|chr1</td>\n",
       "      <td>ENST00000370459</td>\n",
       "      <td>GBP5</td>\n",
       "      <td>TTACTCTAAAGTGCTAAATATGG</td>\n",
       "      <td>ATGGCTTTAGAGATCCACATGTCAGACCCCATGTGCCTCATCGAGA...</td>\n",
       "      <td>GCTAATTGTTGTAGATCATCACTTCAAGGTGCCCATATCTTTCTAG...</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCDS1037.1|Hs109|chr1</td>\n",
       "      <td>ENST00000368737</td>\n",
       "      <td>S100A12</td>\n",
       "      <td>TTACCACACCCACAAAGAGTAGG</td>\n",
       "      <td>ATGACAAAACTTGAAGAGCATCTGGAGGGAATTGTCAATATCTTCC...</td>\n",
       "      <td>CTTCCTTGGCTCAGTGCCCTTCACCACTGCTGGCTTTTTGCTGTAG...</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CCDS1318.1|Hs109|chr1</td>\n",
       "      <td>ENST00000367674</td>\n",
       "      <td>TNR</td>\n",
       "      <td>TTACAGTTCTGAGCAGTGGGCGG</td>\n",
       "      <td>ATGGGGGCAGATGGGGAAACAGTGGTTCTGAAGAACATGCTCATTG...</td>\n",
       "      <td>ACGTCAGAGGCAGGAACCGACTGTGCTAAGGCTGTTGGCTCAACAC...</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CCDS1997.1|Hs109|chr2</td>\n",
       "      <td>ENST00000390655</td>\n",
       "      <td>CD8B</td>\n",
       "      <td>TTACAAATGAGCAGAGAATACGG</td>\n",
       "      <td>ATGCGGCCGCGGCTGTGGCTCCTCTTGGCCGCGCAGCTGACAGTTC...</td>\n",
       "      <td>AGGTGTCCCGGGCGCGCCACGATGCGGCCGCGGCTGTGGCTCCTCT...</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>CCDS82410.1|Hs109|chr19</td>\n",
       "      <td>ENST00000425570</td>\n",
       "      <td>ZNF418</td>\n",
       "      <td>TTACAAGTGAAAAGAAATTTGGG</td>\n",
       "      <td>ATGCAAGAAGGCCTTCATAGAATGACTTGGGACTGCATTGCTATTG...</td>\n",
       "      <td>CTCTGGTAGCGACCATTTTGGTTAATGTTGGGTGTGTTTCTGCGGT...</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>CCDS82506.1|Hs109|chr2</td>\n",
       "      <td>ENST00000452780</td>\n",
       "      <td>EPB41L5</td>\n",
       "      <td>TTACTGACCACTGAGCTCTGAGG</td>\n",
       "      <td>ATGCTGAGTTTCTTCCGTAGAACACTAGGGCGTCGGTCTATGCGTA...</td>\n",
       "      <td>ATGCTGAGTTTCTTCCGTAGAACACTAGGGCGTCGGTCTATGCGTA...</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>CCDS82829.1|Hs109|chr3</td>\n",
       "      <td>ENST00000460856</td>\n",
       "      <td>KALRN</td>\n",
       "      <td>TTACGTTTAGCGCGCATCCTGGG</td>\n",
       "      <td>ATGACGGACCGCTTCTGGGACCAGTGGTATCTCTGGTATCTCCGCT...</td>\n",
       "      <td>GTGTGGGAGGACTGGTGGCTGTCTTTGCAGGCAGGCATTTGCTTAG...</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>CCDS86415.1|Hs109|chr14</td>\n",
       "      <td>ENST00000556285</td>\n",
       "      <td>TGFB3</td>\n",
       "      <td>TTACTGCTTCCGGTGAGACTGGG</td>\n",
       "      <td>ATGAAGATGCACTTGCAAAGGGCTCTGGTGGTCCTGGCCCTGCTGA...</td>\n",
       "      <td>GGGGAGGCCGCCTGGTTTTCCTCCCTCCTTCTGCACGTCTGCTGGG...</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>CCDS86870.1|Hs109|chr2</td>\n",
       "      <td>ENST00000643224</td>\n",
       "      <td>CD8B2</td>\n",
       "      <td>TTACAAATGAGCAGAGAATACGG</td>\n",
       "      <td>ATGCGGCCGCGGCTGTGGCTCCTCCTGGCCGCGCAGCTGACAGTTC...</td>\n",
       "      <td>ACCCACCCCAGCCGCGACTGTCTCCGCCGAGCCCCCGGGGCCAGGT...</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  CCDS Name        ENST Name Gene Name  \\\n",
       "0      CCDS503.1|Hs109|chr1  ENST00000372343     IPO13   \n",
       "1      CCDS722.1|Hs109|chr1  ENST00000370459      GBP5   \n",
       "2     CCDS1037.1|Hs109|chr1  ENST00000368737   S100A12   \n",
       "3     CCDS1318.1|Hs109|chr1  ENST00000367674       TNR   \n",
       "4     CCDS1997.1|Hs109|chr2  ENST00000390655      CD8B   \n",
       "..                      ...              ...       ...   \n",
       "71  CCDS82410.1|Hs109|chr19  ENST00000425570    ZNF418   \n",
       "72   CCDS82506.1|Hs109|chr2  ENST00000452780   EPB41L5   \n",
       "73   CCDS82829.1|Hs109|chr3  ENST00000460856     KALRN   \n",
       "74  CCDS86415.1|Hs109|chr14  ENST00000556285     TGFB3   \n",
       "75   CCDS86870.1|Hs109|chr2  ENST00000643224     CD8B2   \n",
       "\n",
       "                       gRNA  \\\n",
       "0   TTACACAGCTGACTACTGAGGGG   \n",
       "1   TTACTCTAAAGTGCTAAATATGG   \n",
       "2   TTACCACACCCACAAAGAGTAGG   \n",
       "3   TTACAGTTCTGAGCAGTGGGCGG   \n",
       "4   TTACAAATGAGCAGAGAATACGG   \n",
       "..                      ...   \n",
       "71  TTACAAGTGAAAAGAAATTTGGG   \n",
       "72  TTACTGACCACTGAGCTCTGAGG   \n",
       "73  TTACGTTTAGCGCGCATCCTGGG   \n",
       "74  TTACTGCTTCCGGTGAGACTGGG   \n",
       "75  TTACAAATGAGCAGAGAATACGG   \n",
       "\n",
       "                                        CCDS Sequence  \\\n",
       "0   ATGGAGCGGCGGGAGGAGCAGCCGGGGGCTGCAGGGGCTGGAGCAG...   \n",
       "1   ATGGCTTTAGAGATCCACATGTCAGACCCCATGTGCCTCATCGAGA...   \n",
       "2   ATGACAAAACTTGAAGAGCATCTGGAGGGAATTGTCAATATCTTCC...   \n",
       "3   ATGGGGGCAGATGGGGAAACAGTGGTTCTGAAGAACATGCTCATTG...   \n",
       "4   ATGCGGCCGCGGCTGTGGCTCCTCTTGGCCGCGCAGCTGACAGTTC...   \n",
       "..                                                ...   \n",
       "71  ATGCAAGAAGGCCTTCATAGAATGACTTGGGACTGCATTGCTATTG...   \n",
       "72  ATGCTGAGTTTCTTCCGTAGAACACTAGGGCGTCGGTCTATGCGTA...   \n",
       "73  ATGACGGACCGCTTCTGGGACCAGTGGTATCTCTGGTATCTCCGCT...   \n",
       "74  ATGAAGATGCACTTGCAAAGGGCTCTGGTGGTCCTGGCCCTGCTGA...   \n",
       "75  ATGCGGCCGCGGCTGTGGCTCCTCCTGGCCGCGCAGCTGACAGTTC...   \n",
       "\n",
       "                                        ENST Sequence  Frame  \\\n",
       "0   GCTTGTCTTGTCAGTCACTGGGGCGGAGGCAGCGGCTGTAGCGGGG...      3   \n",
       "1   GCTAATTGTTGTAGATCATCACTTCAAGGTGCCCATATCTTTCTAG...      1   \n",
       "2   CTTCCTTGGCTCAGTGCCCTTCACCACTGCTGGCTTTTTGCTGTAG...      3   \n",
       "3   ACGTCAGAGGCAGGAACCGACTGTGCTAAGGCTGTTGGCTCAACAC...      1   \n",
       "4   AGGTGTCCCGGGCGCGCCACGATGCGGCCGCGGCTGTGGCTCCTCT...      3   \n",
       "..                                                ...    ...   \n",
       "71  CTCTGGTAGCGACCATTTTGGTTAATGTTGGGTGTGTTTCTGCGGT...      3   \n",
       "72  ATGCTGAGTTTCTTCCGTAGAACACTAGGGCGTCGGTCTATGCGTA...      1   \n",
       "73  GTGTGGGAGGACTGGTGGCTGTCTTTGCAGGCAGGCATTTGCTTAG...      3   \n",
       "74  GGGGAGGCCGCCTGGTTTTCCTCCCTCCTTCTGCACGTCTGCTGGG...      3   \n",
       "75  ACCCACCCCAGCCGCGACTGTCTCCGCCGAGCCCCCGGGGCCAGGT...      3   \n",
       "\n",
       "    Distance from 3' End  \n",
       "0                     15  \n",
       "1                      5  \n",
       "2                     18  \n",
       "3                      8  \n",
       "4                      6  \n",
       "..                   ...  \n",
       "71                     6  \n",
       "72                    17  \n",
       "73                     6  \n",
       "74                    12  \n",
       "75                     6  \n",
       "\n",
       "[76 rows x 8 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Search in the forward direction for SpCas9 gRNAs that overlap with stop/5' UTR, TTAC at position 20, any frame\n",
    "#Part 2, with ENST gRNAs\n",
    "\n",
    "#Parameters:\n",
    "search_regex = \"(TTAC\\w{0,17}$)|(TTAC\\w{17}G$)\" #Regex for sequence to match\n",
    "file_name = \"5_UTR_Overlapping_Forward_SpCas9_Filtered\"\n",
    "\n",
    "#Initialize relevant variables\n",
    "counter = 0\n",
    "good_sequences = []\n",
    "filtered_data = []\n",
    "\n",
    "#Load sequences\n",
    "pbar = tqdm(initial_filtered_sequences)\n",
    "\n",
    "#Iterate through sequences and search for matches\n",
    "for sequence in pbar: \n",
    "    #Extract CCDS ID and find corresponding ENST ID\n",
    "    current_CCDS = re.match(\"(CCDS.*)\\.\\d\", sequence.name).group(1)\n",
    "    current_ENST_name = CCDStoENST.get(current_CCDS)\n",
    "    if current_ENST_name:\n",
    "        current_ENST = good_ENST_sequences.get(current_ENST_name) #Dictionary lookup for ENST sequence\n",
    "        if current_ENST is not None:\n",
    "            #Find correct match from before in CCDS\n",
    "            last50 = str(sequence.seq)[-50:] #If you don't truncate, regex search totally kills your memory :(\n",
    "            target_RC_search = re.finditer(search_regex, last50) #Regex match iterator\n",
    "            for match in target_RC_search:\n",
    "                (start, stop) = match.span()\n",
    "                ENST_seq_to_match = last50[:start] #Set TTAC-preceeding sequence to find in ENST\n",
    "                ENST_match = re.finditer(ENST_seq_to_match, str(current_ENST.seq))\n",
    "                for possible_seq in ENST_match:\n",
    "                    (ENST_ignore, ENST_start) = possible_seq.span()\n",
    "                    if str(current_ENST.seq)[ENST_start:ENST_start+4] == \"TTAC\": #Check that is TTAC site\n",
    "                        if str(current_ENST.seq)[ENST_start+21:ENST_start+23] == \"GG\": #Check for PAM\n",
    "                            gRNA_sequence = str(current_ENST.seq)[ENST_start:ENST_start+23]\n",
    "                            frame = (start+4)%3 + 1\n",
    "                            gRNA_dist = 50-start-4\n",
    "                            gene_name = gene_dict[re.match(\"CCDS.*\\.\\d\", sequence.name).group()]\n",
    "                            good_sequences.append(current_ENST)\n",
    "                            filtered_data.append([sequence.name, current_ENST_name, gene_name, gRNA_sequence, str(sequence.seq), str(current_ENST.seq), frame, gRNA_dist])\n",
    "                            counter += 1\n",
    "        \n",
    "    \n",
    "print(f\"Number of matching sequences: {counter}\")\n",
    "\n",
    "#Write data to relevant files\n",
    "with open(file_name+\"_ENST_Matches.fasta\", \"w\") as output_handle:\n",
    "        SeqIO.write(good_sequences, output_handle, \"fasta\")\n",
    "        \n",
    "df = pd.DataFrame(filtered_data, columns = [\"CCDS Name\", \"ENST Name\", \"Gene Name\", \"gRNA\", \"CCDS Sequence\", \"ENST Sequence\", \"Frame\", \"Distance from 3' End\"])\n",
    "df.to_csv(file_name+\".csv\")\n",
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
